File: js/stack/503nextGreaterElements.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var nextGreaterElements = function(nums) {
	let stack = [];
	let ans = [];
	let count = 2;
	while(count>0){
		for (let i = nums.length-1; i >= 0; i--) {
			while(nums[stack[stack.length-1]] <= nums[i]){
				stack.pop();
			}
			if(stack.length>0){
				ans[i] = nums[stack[stack.length-1]];
			}else{
				ans[i] = -1;
			}
			stack.push(i);
		}
		count--;		
	}
	return ans;
};

let testcase = nextGreaterElements([-1,0]);
console.log(testcase);
export default nextGreaterElements;

File: py/stack/503nextGreaterElements.py
from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        db_nums = nums*2
        ans = [-1]*len(nums)
        stack = [0]
        len_db = len(db_nums)-1
        for i in range(len_db,-1,-1):
            while len(stack) >1 and db_nums[i] >= nums[stack[-1]]:
                stack.pop()
            if len(stack) >1:
                top_idx = stack[-1]
                tmp = int(i%len(nums))
                ans[tmp] = nums[top_idx]
            stack.append(i % len(nums))
        return ans



solution = Solution()
ans = solution.nextGreaterElements([3,4,2,1,6,7,-1,-5,10,-9])
print(ans)

